whylog/config/__init__.py

```diff
@@ -53,57 +60,81 @@ def _save_rule_definition(self, rule_definition):
-            '^(\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d) visible effect$', [1], 'default', {1: 'date'}
+    def create_investigation_plan(self, front_input, log_type):
+        matching_parsers, effect_params = self._find_matching_parsers(
+            front_input.line_content, log_type.name
         )
-        concatenated = ConcatenatedRegexParser([cause])
+        suspected_rules = self._filter_rule_set(matching_parsers)
+        concatenated_parsers = self._create_concatenated_parsers_for_investigation(suspected_rules)
+        #TODO: creating clues base on effect_params
+        #TODO: remove mocks
         effect_time = datetime(2015, 12, 3, 12, 8, 9)
-        earliest_cause_time = datetime(2015, 12, 3, 12, 8, 8)
-        default_investigation_step = InvestigationStep(
-            concatenated, effect_time, earliest_cause_time
-        )
-        rule = Rule(
-            [cause], effect, [
-                {
-                    'clues_groups': [[1, 1], [0, 1]],
-                    'name': 'time',
-                    'params': {'max_delta': 1}
-                }
-            ]
-        )  # yapf: disable
         line_source = LineSource('localhost', 'node_1.log', 40)
         effect_clues = {'effect': Clue((effect_time,), 'node_1.log', line_source)}
-        return InvestigationPlan(
-            [rule], [(default_investigation_step, default_log_type)], effect_clues
+        steps = self._create_steps_in_investigation(
+            concatenated_parsers, suspected_rules, effect_clues
         )
-
-    def create_investigation_plan(self, front_input, log_type_name):
-        # TODO: remove mock
-        return self.mocked_investigation_plan()
+        return InvestigationPlan(suspected_rules, steps, effect_clues)
 
     def get_log_type(self, front_input):
         # TODO: remove mock
         matcher = RegexFilenameMatcher('localhost', 'node_1.log', 'default')
         return LogType('default', [matcher])
 
-    def _find_matching_parsers(self, front_input, log_type):
-        pass
+    def _find_matching_parsers(self, effect_line_content, log_type_name):
+        """
+        This method finding all parsers from Config base which matching with effect_line_content
+        """
+        matching_parsers = []
+        extracted_params = {}
+        for parser in self._parsers_grouped_by_log_type[log_type_name]:
+            params = parser.get_regex_params(effect_line_content)
+            if params is not None:
+                extracted_params[parser.name] = params
+                matching_parsers.append(parser)
+        return matching_parsers, extracted_params
 
     def _filter_rule_set(self, parsers_list):
-        pass
+        """
+        This method finding all rules from Config base which can be fulfilled in
+        single investigation base on parsers_list found by _find_matching_parsers
+        """
+        suspected_rules = []
+        for parser in parsers_list:
+            if self._rules.get(parser.name) is not None:
```

Comment:
Assign to a temporary variable instead of calling `.get()` twice. The whole point of using get instead of `in` operator is to avoid doing a lookup in the underlying structure twice.